inference-engine/src/gna_plugin/transformations/insert_reshape_around_matmul.cpp

```diff
@@ -41,38 +41,56 @@ static bool InsertReshape(
         return false;
     }
 
-    std::shared_ptr<ngraph::Node> first_node = iter->second.get_node_shared_ptr();
+    auto first_node = iter->second.get_node_shared_ptr();
+    std::vector<std::shared_ptr<ngraph::Node>> nodes = { matmul_node };
+    for (auto node : {add2, add1, fake_quantize, transpose}) {
+        iter = pattern_map.find(node);
+        if (iter != pattern_map.end()) {
+            nodes.push_back(iter->second.get_node_shared_ptr());
+        }
+    }
+
+    auto last_node_shape = nodes.back()->get_output_shape(0);
     auto reshape_input_node = std::dynamic_pointer_cast<ngraph::opset8::Reshape>(first_node);
     bool need_reshape_before = !reshape_input_node || reshape_input_node->get_output_shape(0).size() != 2;
     if (need_reshape_before) {
         auto input_shape = first_node->get_output_shape(0);
-        std::vector<size_t> before_shape(2, 1);
-        std::copy_if(input_shape.begin(), input_shape.end(), before_shape.begin(), [](size_t e) { return e > 1; });
+        std::vector<int> before_shape(2, -1);
+        before_shape[1] = input_shape.back();
         auto reshape_before_node = std::make_shared<ngraph::opset8::Reshape>(first_node,
             std::make_shared<ngraph::opset8::Constant>(ngraph::element::Type_t::i64, ngraph::Shape{before_shape.size()}, before_shape), false);
         reshape_before_node->set_friendly_name(matmul_node->get_friendly_name() + "/reshape_before_matmul");
         ngraph::copy_runtime_info(first_node, reshape_before_node);
         matmul_node->input(matmul_input_index).replace_source_output(reshape_before_node->output(0));
-    }
+        if (auto transpose_node = std::dynamic_pointer_cast<ngraph::opset8::Transpose>(nodes.back())) {
+            nodes.pop_back();
+            std::reverse(nodes.begin(), nodes.end());
+            while (!nodes.empty()) {
+                auto node_copy = nodes.back()->clone_with_new_inputs(nodes.back()->input_values());
+                ngraph::copy_runtime_info(nodes.back(), node_copy);
+                ngraph::replace_node(nodes.back(), node_copy);
+                nodes.pop_back();
+            }
 
-    std::shared_ptr<ngraph::Node> last_node;
-    iter = pattern_map.find(transpose);
-    if (iter == pattern_map.end() &&
-        (iter = pattern_map.find(fake_quantize2)) == pattern_map.end() &&
-        (iter = pattern_map.find(add1)) == pattern_map.end() &&
-        (iter = pattern_map.find(add2)) == pattern_map.end()) {
-        last_node = matmul_node;
-    } else {
-        last_node = iter->second.get_node_shared_ptr();
+            auto transpose_input_shape = transpose_node->input_values()[0].get_node_shared_ptr()->get_output_shape(0);
+            auto transpose_constant_shape = transpose_node->input_values()[1].get_node_shared_ptr()->get_output_shape(0);
+            IE_ASSERT((std::count_if(transpose_input_shape.begin(), transpose_input_shape.end(), [](size_t n) { return n > 1; }) <= 2));
```

Comment:
it's better to throw exception with some more clear explanation, since this situation theoretically can occur in the customer's network

Reply:
done